matrix_sdk/src/room/common.rs

```diff
@@ -0,0 +1,104 @@
+    message::get_message_events,
+};
+use std::ops::Deref;
+
+use crate::{Client, Result, Room, RoomMember};
+
+/// A struct containing methodes that are common for Joined, Invited and Left Rooms
+#[derive(Debug, Clone)]
+pub struct Common {
+    inner: Room,
+    pub(crate) client: Client,
+}
+
+impl Deref for Common {
+    type Target = Room;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Common {
+    /// Create a new `room::Common`
+    ///
+    /// # Arguments
+    /// * `client` - The client used to make requests.
+    ///
+    /// * `room` - The underlaying room.
+    pub fn new(client: Client, room: Room) -> Self {
+        Self {
+            inner: room,
+            client,
+        }
+    }
+
+    /// Forget this room.
+    ///
+    /// This communicates to the homeserver that it should forget the room.
+    pub async fn forget(&self) -> Result<()> {
+        let request = forget_room::Request::new(self.inner.room_id());
+        let _response = self.client.send(request, None).await?;
+
+        Ok(())
+    }
+
+    /// Leave this room.
+    ///
+    /// Only invited and joined rooms can be left
+    pub(crate) async fn leave(&self) -> Result<()> {
+        let request = leave_room::Request::new(self.inner.room_id());
+        let _response = self.client.send(request, None).await?;
+
+        Ok(())
+    }
+
+    /// Join this room.
+    ///
+    /// Only invited and left rooms can be joined via this method
+    pub(crate) async fn join(&self) -> Result<()> {
+        let request = join_room_by_id::Request::new(self.inner.room_id());
+        let _resposne = self.client.send(request, None).await?;
+
+        Ok(())
+    }
+
+    /// Sends a request to `/_matrix/client/r0/rooms/{room_id}/messages` and returns
+    /// a `get_message_events::Response` that contains a chunk of room and state events
+    /// (`AnyRoomEvent` and `AnyStateEvent`).
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - The easiest way to create this request is using the
+    /// `get_message_events::Request` itself.
+    pub async fn messages(
+        &self,
+        request: impl Into<get_message_events::Request<'_>>,
+    ) -> Result<get_message_events::Response> {
+        let request = request.into();
+        self.client.send(request, None).await
+    }
+
+    pub(crate) async fn request_members(&self) -> Result<()> {
+        let request = get_member_events::Request::new(self.inner.room_id());
+        let response = self.client.send(request, None).await?;
+
+        self.client
+            .base_client
+            .receive_members(self.inner.room_id(), &response)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Get the room active members for the given room, includes invited and joined members.
+    pub async fn members(&self) -> Result<Vec<RoomMember>> {
+        // TODO: get all members, active_memebers doesn't include left members
```

Comment:
Please call this method then `get_active_members()` and add a method that returns all members separately. Some clients will want to operate only on the joined/invited subset.

I was under the impression that the store can return all members but it seems that I forgot about it. To add support for this you'll have to add a method to the trait and in the sled store you'll have to add a modified version of [this](https://github.com/matrix-org/matrix-rust-sdk/blob/master/matrix_sdk_base/src/store/sled_store/mod.rs#L525) method, you just replace the `get((room_id.as_str(), state_key.as_str()).encode())` call with an `scan_prefix(room_id.encode())` call.